### ISSUE TITLE
flake8 --output-file no longer generates output file as expected.

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#725]
In GitLab by @bnp26 on May 11, 2020, 17:37

Please read this brief portion of documentation before going any further: http://flake8.pycqa.org/en/latest/internal/contributing.html#filing-a-bug

*Please describe how you installed Flake8*
```
$ pip install flake8
```
**Note**: Some *nix distributions patch Flake8 arbitrarily to accommodate incompatible software versions. If you're on one of those distributions, your issue may be closed and you will be asked to open an issue with your distribution package maintainers instead.

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```{
  "dependencies": [],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.7.7",
    "system": "Darwin"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.6.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.2.0"
    }
  ],
  "version": "3.8.0"
}
```

*Please describe the problem or feature*

```flake8 does not output to the output-file as it did before v3.8.0
doesn't work with it being specified in both the config as well as in the cmd itself.
```
*If this is a bug report, please explain with examples (and example code) what you expected to happen and what actually happened.*
```
flake8 hello-world.py --output-file=/tmp/flake8.log

ouput:
hello-world.py:1:1: F401 're' imported but unused
hello-world.py:3:1: E302 expected 2 blank lines, found 1
hello-world.py:6:1: E302 expected 2 blank lines, found 1
hello-world.py:9:1: E305 expected 2 blank lines after class or function definition, found 1

expected output:
```

**Comment by asottile** [comment]
In GitLab by @bnp26 on May 11, 2020, 17:38

changed the description

**Comment by asottile** [comment]
In GitLab by @bnp26 on May 11, 2020, 17:39

changed the description

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 17:54

this indeed appears to be a regression in 3.8 -- can you work around with `>` for now, I will try and find a fix for this (a lot of the argument parsing was reorganized between 3.7 and 3.8 to make it easier to add features)

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 18:09

CC @ericvw in case you have any quick thoughts on this one -- looks like some of the preliminary options are being used afterwards but not forwarded along -- a quick fix would be to merge the two

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 18:34

mentioned in merge request !426

**Comment by asottile** [comment]
In GitLab by @ericvw on May 11, 2020, 19:54

@asottile This was removed from being parsed from the configuration on purpose… see !361.

> The `--output-file` option is used when parsing and handling preliminary options for logging. It is also used to set `BaseFormatter.filename`. Towards ensuring that the option is respected consistently for both logging and formatting plugins, only observe the value from the CLI.

Otherwise, one would get split output if two different files were specified on the CLI and in the configuration file.  This was the awkwardness I was attempting to "resolve" in !361.

The intent of the CLI rework was to avoid forwarding preliminary and make a clear distinction (see !364).  I was purposely avoiding parsing `argv` twice, which is what I see in !4261.

I need to investigate further, but if no logging occurs after the preliminary parsing of the CLI, the logging configuration may be deferred to after parsing configuration.

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 19:57

:thumbsup: makes sense -- the problem now is that `BaseFormatter.filename` is always `None` (since by the time the final argparser parses arguments, `options.output_file` is always `None`)

**Comment by asottile** [comment]
In GitLab by @ericvw on May 11, 2020, 20:10

I'm still getting my head wrapped around the `BaseFormatter`, but I am curious to explore if `output_file` could be forwarded to the `BaseFormatter.__init__(self, options)` method to be `BaseFormatter.__init__(self, options, output_file)`.  I have no idea the implications of this yet because I haven't fully understood this part of the code base… yet.

**Comment by asottile** [comment]
In GitLab by @ericvw on May 11, 2020, 20:18

If I am reading things correctly, it looks like `BaseFormatter` only needs 3 options from `options` argument passed during its initialization.

How about changing the signature of `BaseFormater.__init__` to something like `def __init__(self, filename, tee, show_source)`?

**Comment by asottile** [comment]
In GitLab by @ericvw on May 11, 2020, 20:19

Scratch that — I didn't realize `BaseFormatter` was a public-facing for developing formating plugins.

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 20:25

yeah our hands are a bit tied with the interface there sadly :(

**Comment by asottile** [comment]
In GitLab by @ericvw on May 11, 2020, 20:34

I think I have a simpler "fix" to special-case the output file handling.  Merge request on its way soon.

**Comment by asottile** [comment]
In GitLab by @ericvw on May 11, 2020, 20:50

mentioned in merge request !427

**Comment by asottile** [comment]
In GitLab by @ericvw on May 11, 2020, 20:52

@asottile I created !427, reusing your test case :smile:.  It's a less invasive change that should also work.

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 22:21

closed via merge request !427

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 22:21

mentioned in commit 00985a64bce799583d438bb3893acfd8a08fde93

**Comment by asottile** [comment]
In GitLab by @asottile on May 11, 2020, 22:36

this has been released as part of 3.8.1 -- thanks again for the report!
